sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuSubPartitionHashJoin.scala

```diff
@@ -166,46 +183,122 @@ class GpuBatchSubPartitioner(
+        val gpuBatch = batchesIter.next()
+        if (gpuBatch.numRows() > 0 && gpuBatch.numCols() > 0) {
+          val types = closeOnExcept(gpuBatch)(GpuColumnVector.extractTypes)
+          // 1) Hash partition on the batch
+          val partedTable = GpuHashPartitioningBase.hashPartitionAndClose(gpuBatch,
+            inputBoundKeys, numParts, "Sub-Hash Calculate", hashSeed)
+          // 2) Split into smaller tables according to partitions
+          val subTables = withResource(partedTable) { _ =>
+            partedTable.getTable.contiguousSplit(partedTable.getPartitions.tail: _*)
+          }
+          // 3) Make each smaller table spillable and cache them in the queue
+          withResource(subTables) { _ =>
+            subTables.zipWithIndex.foreach { case (table, id) =>
+              // skip empty tables
+              if (table.getRowCount > 0) {
+                subParts(id) += SpillableColumnarBatch(table, types,
+                  SpillPriorities.ACTIVE_ON_DECK_PRIORITY)
+              }
             }
           }
+        } else if (gpuBatch.numRows() > 0 && gpuBatch.numCols() == 0) {
+          // Rows only batch. This should never happen for a hash join in Spark.
+          gpuBatch.close()
+        } else {
+          // Skip empty batches
+          gpuBatch.close()
         }
-      } else if (gpuBatch.numRows() > 0 && gpuBatch.numCols() == 0) {
-        // Rows only batch. This should never happen for a hash join in Spark.
-        gpuBatch.close()
-      } else {
-        // Skip empty batches
-        gpuBatch.close()
       }
-    } // end of while
+      subParts
+    }
+  }
+
+  override final def currentSeed: Int = seedGenerator.currentSeed
+
+  /**
+   * Always return the current seed.
+   * This is intended to share the same seed across sub-partitioners.
+   */
+  override final def nextSeed: Int = seedGenerator.currentSeed
+}
+
+/**
+ * A sub-partitioner supports partitioning the input batches multiple times to try to
+ * have each sub partition been smaller than the target batch size.
+ * However it may fail if the data is highly skewed, meaning there may be some
+ * partitions still going beyond the target batch size.
+ */
+class GpuBatchSizeAwareSubPartitioner(
+    inputIter: Iterator[ColumnarBatch],
+    inputBoundKeys: Seq[GpuExpression],
+    numPartitions: Int,
+    targetBatchSize: Long,
+    seedGenerator: SeedGenerator = new SimpleSeedGenerator())
+  extends GpuBatchSubPartitioner(inputIter, inputBoundKeys, numPartitions, seedGenerator) {
+
+  private var retryCount = 0
+  private var actualNumPartitions = 0
+
+  private[this] def computeNumPartitions(parts: SubPartitionBuffer): Int = {
+    val totalSize = parts.flatten.map(_.sizeInBytes).sum
+    Math.floorDiv(totalSize, math.max(targetBatchSize, 1)).toInt + 1
+  }
+
+  private[this] def needRepartition(parts: SubPartitionBuffer): Boolean = {
+    // FIXME Is it good enough to ask for repartitioning when there exists any sub
```

Comment:
I think we would want to stop for a given batch if we try to partition the batch and it didn't change. Meaning we tried to partition it N ways and we got back N-1 empty batches and 1 batch with all of the data in it.

That said I think we only need to do two passes through the data at most.  The algorithm I see would be something like the following.

* If the build side < target batch size
    * do the join as normal
* else
    * partition the build/stream sides according to the configured number of splits
    * group build partitions to try and be just under target batch size and do the join for all build partitions that met this goal
    * for each build partition > target batch size 
        * num_partitions = ceil(build size / target batch size)
        * partition the build and stream batches using the new seed and the calculated number of partitions
        * group partitions together again like before and do the join no matter how large the size is.

Reply:
Done.